drivers/kscan/kscan_mchp_xec.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <drivers/kscan.h>
+#include <kernel.h>
+#include <soc.h>
+#include <sys/atomic.h>
+#include <logging/log.h>
+
+#define LOG_LEVEL CONFIG_KSCAN_LOG_LEVEL
+LOG_MODULE_REGISTER(kscan_mchp_xec);
+
+#define MAX_MATRIX_KEY_COLS CONFIG_KSCAN_XEC_COLUMN_SIZE
+#define MAX_MATRIX_KEY_ROWS CONFIG_KSCAN_XEC_ROW_SIZE
+
+#define KEYBOARD_COLUMN_DRIVE_ALL       -2
+#define KEYBOARD_COLUMN_DRIVE_NONE      -1
+
+/* Poll period/debouncing rely onthe 32KHz clock with 30 usec clock cycles */
+#define CLOCK_32K_HW_CYCLES_TO_US(X) \
+	(u32_t)((((u64_t)(X) * 1000000U) / sys_clock_hw_cycles_per_sec()))
+/* Milliseconds in microseconds */
+#define MSEC 1000U
+/* Number of tracked scan times */
+#define SCAN_OCURRENCES 30U
+/* Thread stack size */
+#define TASK_STACK_SIZE 1024
+
+struct kscan_xec_data {
+	/* variables in usec units */
+	u32_t deb_time_press;
+	u32_t deb_time_rel;
+	s64_t poll_timeout;
+	u32_t poll_period;
+	u8_t matrix_stable_state[MAX_MATRIX_KEY_COLS];
+	u8_t matrix_unstable_state[MAX_MATRIX_KEY_COLS];
+	u8_t matrix_previous_state[MAX_MATRIX_KEY_COLS];
+	/* Index in to the scan_clock_cycle to indicate start of debouncing */
+	u8_t scan_cycle_idx[MAX_MATRIX_KEY_COLS][MAX_MATRIX_KEY_ROWS];
+	/* Track previous "elapsed clock cycles" per matrix scan. This
+	 * is used to calculate the debouncing time for every key
+	 */
+	u8_t scan_clk_cycle[SCAN_OCURRENCES];
+	struct k_sem poll_lock;
+	u8_t scan_cycles_idx;
+	kscan_callback_t callback_isr;
+	struct k_thread thread;
+	atomic_t enable_scan;
+
+	K_THREAD_STACK_MEMBER(thread_stack, TASK_STACK_SIZE);
+};
+
+static KSCAN_Type *base = (KSCAN_Type *)
+	(DT_INST_0_MICROCHIP_XEC_KSCAN_BASE_ADDRESS);
+
+static struct kscan_xec_data kbd_data;
+
+static void drive_keyboard_column(int data)
+{
+	if (data == KEYBOARD_COLUMN_DRIVE_ALL) {
+		/* KSO output controlled by the KSO_SELECT field */
+		base->KSO_SEL = MCHP_KSCAN_KSO_ALL;
+	} else if (data == KEYBOARD_COLUMN_DRIVE_NONE) {
+		/* Keyboard scan disabled. All KSO output buffers disabled */
+		base->KSO_SEL = MCHP_KSCAN_KSO_EN;
+	} else {
+		/* It is assumed, bit 5 was previously set */
```

Comment:
Document that bit 5 corresponds KEYBOARD_COLUMN_DRIVE_ALL

Reply:
done